source/fuzz/transformation_composite_insert.cpp

```diff
@@ -0,0 +1,209 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "transformation_composite_insert.h"
+
+#include "source/fuzz/fuzzer_pass_add_composite_inserts.h"
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationCompositeInsert::TransformationCompositeInsert(
+    const spvtools::fuzz::protobufs::TransformationCompositeInsert& message)
+    : message_(message) {}
+
+TransformationCompositeInsert::TransformationCompositeInsert(
+    const protobufs::InstructionDescriptor& instruction_to_insert_before,
+    uint32_t fresh_id, uint32_t composite_id, uint32_t object_id,
+    const std::vector<uint32_t>& index) {
+  *message_.mutable_instruction_to_insert_before() =
+      instruction_to_insert_before;
+  message_.set_fresh_id(fresh_id);
+  message_.set_composite_id(composite_id);
+  message_.set_object_id(object_id);
+  for (auto an_index : index) {
+    message_.add_index(an_index);
+  }
+}
+
+bool TransformationCompositeInsert::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |message_.fresh_id| must be fresh.
+  if (!fuzzerutil::IsFreshId(ir_context, message_.fresh_id())) {
+    return false;
+  }
+
+  // |message_.composite_id| must refer to an existing composite value.
+  auto composite =
+      ir_context->get_def_use_mgr()->GetDef(message_.composite_id());
+
+  if (!IsCompositeInstructionSupported(ir_context, composite)) {
+    return false;
+  }
+
+  // |message_.index| must refer to a valid index. The type id of the object at
+  // |message_.object_id| and the type id of the component of the composite at
+  // index |message_.index| must be the same.
+  auto component_to_be_replaced_type_id = fuzzerutil::WalkCompositeTypeIndices(
+      ir_context, composite->type_id(), message_.index());
+  if (component_to_be_replaced_type_id == 0) {
+    return false;
+  }
+  auto object_instruction =
+      ir_context->get_def_use_mgr()->GetDef(message_.object_id());
+  if (object_instruction == nullptr) {
+    return false;
+  }
+
+  // We ignore pointers for now. Consider adding support for pointer types.
+  auto object_instruction_type =
+      ir_context->get_type_mgr()->GetType(object_instruction->type_id());
+  // No unused variables in release mode (to keep compilers happy).
+  (void)object_instruction_type;
+  assert(object_instruction_type->kind() != opt::analysis::Type::kPointer &&
+         "The object to be inserted cannot be a pointer.");
```

Comment:
I believe we currently like to imagine that transformations could be applied to ANY module, and they should never crash. This means the `IsApplicable` method should not crash, but may return false (and so the `Apply` method will never be called). Theoretically, if `Apply` was somehow called even though `IsApplicable` would return false, we are allowed to crash (assert false). 

Thus, the instruction found from `message_.object_id` might end up not having a type id. So you should do `if (!object_instruction->type_id()) return false;`.

Similarly, I think the assert should not be an assert, and we should instead just return false if `object_instruction_type` happens to be a pointer type. Also, I think we prefer to use `->AsPointer()` (or `->AsInteger()`, `->AsFloat()`, etc.) to check the type; we will get nullptr if the type is not a pointer (integer, float, etc.).

Reply:
Actually, I guess you technically don't need to check that `object_instruction->type_id()` is non-zero because it is compared with `component_to_be_replaced_type_id`, which is non-zero. But in general, it does not hurt to check this.

Reply:
Indeed. I agree. It was me who requested that we add an assertion here (sorry, @antnkarp).

But now that I think about it, it might indeed be better to use an `if` statement here. I guess the rule of thumb here is:
> if some condition is valid but not supported (like the pointer case above) - use `if`s; otherwise (if the condition is not possible in the current context) - use asserts. 

The last part might seem like a waste of time but it helps the reader to understand writer's assumptions about the code.